lib/AST/DiagnosticEngine.cpp

```diff
@@ -140,6 +143,50 @@ struct EducationalNotes {
 static constexpr EducationalNotes<LocalDiagID::NumDiags> _EducationalNotes = EducationalNotes<LocalDiagID::NumDiags>();
 static constexpr auto educationalNotes = _EducationalNotes.value;
 
+class LocalizationInput : public llvm::yaml::Input {
+  using Input::Input;
+  
+  /// Read diagnostics in the YAML file iteratively
+  template <typename T, typename Context>
+  static
+  typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
+  readYAML(IO &io, T &Seq, bool, Context &Ctx) {
+    unsigned count = io.beginSequence();
+    
+    // Resize Diags from YAML file to be the same size
+    // as diagnosticStrings from def files.
+    Seq.resize(LocalDiagID::NumDiags);
+    for(unsigned i = 0; i < count; ++i) {
+      void *SaveInfo;
+      if (io.preflightElement(i, SaveInfo)) {
+        DiagnosticNode current;
+        yamlize(io, current, true, Ctx);
+        io.postflightElement(SaveInfo);
+        
+        // YAML file isn't guaranteed to have diagnostics in order of their
+        // declaration in `.def` files, to accommodate that we need to leave
+        // holes in diagnostic array for diagnostics which haven't yet been
+        // localized and for the ones that
+        // have `DiagnosticNode::id` indicates their position.
+        Seq[static_cast<unsigned>(current.diag_id)] = std::move(current);
+      }
+    }
+    io.endSequence();
+  }
+
+  template <typename T>
+  inline friend
+  typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value,
+  LocalizationInput &>::type
+  operator>> (LocalizationInput &yin, T &diagnostics) {
+    llvm::yaml::EmptyContext Ctx;
+    if (yin.setCurrentDocument())
+      readYAML(yin, diagnostics, true, Ctx);
+
```

Comment:
I think it would make sense to warn if YAML file couldn't be read, WDYT?

Reply:
I added a comment. Do you think it's sufficient?

Reply:
How would the developer know that provided path is incorrect?